modules/bullet/shape_bullet.cpp

```diff
@@ -438,7 +438,47 @@ void HeightMapShapeBullet::set_data(const Variant &p_data) {
 
 	int l_width = d["width"];
 	int l_depth = d["depth"];
-	PoolVector<real_t> l_heights = d["heights"];
+
+	// TODO This code will need adjustments if real_t is set to `double`,
+	// because that precision is unnecessary for a heightmap and Bullet doesn't support it...
+
+	PoolVector<real_t> l_heights;
+	Variant l_heights_v = d["heights"];
+
+	if (l_heights_v.get_type() == Variant::POOL_REAL_ARRAY) {
+		// Ready-to-use heights can be passed
+
+		l_heights = l_heights_v;
+
+	} else if (l_heights_v.get_type() == Variant::OBJECT) {
+		// If an image is passed, we have to convert it to a format Bullet supports.
+		// this would be expensive to do with a script, so it's nice to have it here.
+
+		Ref<Image> l_image = l_heights_v;
+		ERR_FAIL_COND(l_image.is_null());
+
+		// Float is the only common format between Godot and Bullet that can be used for decent collision.
+		// (Int16 would be nice too but we still don't have it)
+		// We could convert here automatically but it's better to not be intrusive and let the caller do it if necessary.
+		ERR_FAIL_COND(l_image->get_format() != Image::FORMAT_RF);
+
+		PoolByteArray im_data = l_image->get_data();
+
+		l_heights.resize(l_image->get_width() * l_image->get_width());
```

Comment:
Is this right? Should it be l_image->get_width() * l_image->get_height() ?

Reply:
That's a mistake, it should be width * height.